### DISCUSSION TITLE
[Drupal] Ideate on how to keep seemingly independent packages' versions in sync

### DISCUSSION THREAD

**Discussion by mxr576** [composer/composer#12024] [Q&A / Support]
Drupal core split their "recommended" [production](https://packagist.org/packages/drupal/core-recommended) and [non-production (development) dependencies](https://packagist.org/packages/drupal/core-dev) into two meta-packages. They follow the same versioning, they released together, but technically speaking, nothing prevents installed `drupal/core-recommended ` 10.3.0 with `drupal/core-dev` 10.2.1 or vice versa, which can lead to unexpected bugs in some cases.

So the question what are the possibilities in Composer to keep versions in sync of seemingly independent packages?

For example. could plugins "instruct" the dependency resolution mechanism to keep the installed version of the mentioned packages in sync? There is the `PluginEvents::PRE_POOL_CREATE`, but that may not be ideal for ensuring this requirement both when packages are installed or updated on a project.

**Comment by stof** [comment]
Well, `drupal/core-dev` could have a conflict rule for `drupal/core-recommended: !=self.version`, if composer adds support for such `!=self.version` link (which makes sense for `conflict` links, but not really for requirements). However, note that this does not work for now (only `self.version` is supported, not `!=self.version`). With such conflict rule, if both packages are installed, they must be at the same version (but you could still use `drupal/core-dev` alone).

Alternatively, `drupal/core-dev` could have a _requirement_ on `drupal/core-recommended: self.version`, which would work today. However, it would mean that `drupal/core-recommended` would still get applied for any project using `drupal/core-dev` even though it does not depend on `drupal/core-recommended` (I don't know the Drupal ecosystem enough to know whether this would be an issue or no)

**Reply by mxr576** [reply]
Thx @stof that is a good idea, actually the common package is `drupal/core`, so we can use that as an anchor to ensure the version of drupal/core-recommended and drupal/core-dev matches with the version of drupal/core via self.version.

https://git.drupalcode.org/project/drupal/-/merge_requests/8647

**Comment by joachim-n** [comment]
The documentation on d.org recommends to always do `composer update 'drupal/core-*'` when doing updates of Drupal core, and that way both packages get updated to the same version.

Having this enforced somehow would be nice though!

You could act in PluginEvents::PRE_POOL_CREATE, check if both meta-packages are present or being installed, and if so, say that drupal/core-dev has a dependency on drupal/core-recommended of self.version.

**Reply by mxr576** [reply]
>The documentation on d.org recommends to always do composer update 'drupal/core-*' when doing updates of Drupal core, and that way both packages get updated to the same version. 

and that way both packages **MAY** get updated to the same version or **MAY NOT**... That command just instructs Composer to try to update all matching package, but dependency conflicts can still hold back one, more or all matching packages. 

Btw, [original thread](https://drupal.slack.com/archives/C223PR743/p1719331712444049) from Drupal Slack and the relevant part as a screenshot.

![image](https://github.com/composer/composer/assets/1755573/4c87bf76-0b00-485b-9b02-72bbba75fa01)

(And the link on the screenshot: https://github.com/Pronovix/drupal-qa/commit/81e10ebd2d7d84dfc78c635b3016a362d9651133)
